### PR TITLE
[WIP] active mode support

### PR DESCRIPTION
**Pull request by rsichny**
as per https://github.com/pohmelie/aioftp/issues/16

this works, but still is far away from being suitable for merge...

TODO:
- [ ] decide what to do with `passive_server` in case of active mode (`await asyncio.sleep(0)` is stupid)
- [ ] add client-side support
- [ ] add tests
- [ ] release data port if needed when closing passive server on mode switch
- [ ] catch and correctly process `ConnectionRefusedError` for data connection in active mode

### CODE REVIEW

File: aioftp/server.py [hunk 1]
```diff
@@ -1567,8 +1569,96 @@ def get_paths(self, connection, path):
...
+                throttles=connection.command_connection.throttles,
+                timeout=connection.socket_timeout,
+                loop=connection.loop
+            )
+
+        nums = tuple(map(int, str.split(rest, ",")))
+        host = str.join(".", map(str, nums[:4]))
+        port = (nums[4] << 8) | nums[5]
+
+        if not connection.future.passive_server.done():
```
Comment (pohmelie) [inline aioftp/server.py]: Don't get this check. Passive server already closed and removed from connection above.

File: aioftp/server.py [hunk 2]
```diff
@@ -1567,8 +1569,96 @@ def get_paths(self, connection, path):
         return passive_server
 
     @ConnectionConditions(ConnectionConditions.login_required)
+    async def port(self, connection, rest):
+
+        if connection.is_passive:
+
+            if connection.future.data_connection.done():
```
Comment (pohmelie) [inline aioftp/server.py]: This duplicate the code at 1622. So, it can be done once below, I think.

File: aioftp/server.py [hunk 3]
```diff
@@ -1567,8 +1569,96 @@ def get_paths(self, connection, path):
...
+
+            connection.data_connection.close()
+
+        else:
+
+            connection.future.data_connection.cancel()
+
+        del connection.data_connection
+
+        connection['data_connection'] = asyncio.ensure_future(
```
Comment (pohmelie) [inline aioftp/server.py]: At first, just `connection.data_connection = ...`. At second, if I got it right, connection should be made right after STOR/RETR/MLSD/etc. Not, after PORT.

File: aioftp/server.py [hunk 4]
```diff
@@ -1567,8 +1569,96 @@ def get_paths(self, connection, path):
...
+                loop=connection.loop
+            )
+
+        nums = tuple(map(int, str.split(rest, ",")))
+        host = str.join(".", map(str, nums[:4]))
+        port = (nums[4] << 8) | nums[5]
+
+        if not connection.future.passive_server.done():
+
+            connection.passive_server = await asyncio.sleep(0)
```
Comment (pohmelie) [inline aioftp/server.py]: Don't get this line. `await asyncio.sleep(0)` will return `None`.

File: aioftp/server.py (hunk 3 above)
Comment (rsichny) [inline aioftp/server.py]: `connection.data_connection = ...` will need coroutine result and we need to schedule the coroutine, but not execute it until ConnectionCondition(data_connection_made) is checked...

File: aioftp/server.py (hunk 4 above)
Comment (rsichny) [inline aioftp/server.py]: yeah, but will store future which we need into connection.futures.passive_server (and so all ConnectionCondition(passive_server_started) checks will pass)... as i already said in PR comment - this is something i need to think on

File: aioftp/server.py (hunk 1 above)
Comment (rsichny) [inline aioftp/server.py]: we need `connection.futures.passive_server` to pass `ConnectionCondition(passive_server_started)` checks

File: aioftp/server.py (hunk 3 above)
Comment (pohmelie) [inline aioftp/server.py]: True. But `ensure_future` schedules coroutine immediately.

File: aioftp/server.py (hunk 1 above)
Comment (pohmelie) [inline aioftp/server.py]: Oh, I see… I think we need to extend `ConnectionCondition` logic. Cause making fake passive server is not good, I think.

File: aioftp/server.py (hunk 3 above)
Comment (rsichny) [inline aioftp/server.py]: > True. But ensure_future schedules coroutine immediately.

hmm, yeah, my bad.... now i should go and find out why it works correctly (i.e. connection is created only in LIST/STOR/etc) :8ball:

File: aioftp/server.py (hunk 4 above)
Comment (pohmelie) [inline aioftp/server.py]: In this case you can just write `connection.passive_server = None`. But as I said above, we need to change `ConnectionCondition` beahaviour.

**Comment by rsichny** [comment]
I believed that i will have time to work on this but seems like not, closing as it's really outdated.

**Comment by pohmelie** [comment]
Don't worry, since I definitely will rewrite aioftp as sioftp (sans-io ftp) with slightly different architecture there will be place for active mode :wink:

**Comment by ramstein74** [comment]
I really need active mode support.
What is the status of this feature ?

**Comment by pohmelie** [comment]
@ramstein74 
This pull is outdated.

Active mode is rarely used. What is the reason you can't use `EPSV`/`PASV` instead?

**Comment by ramstein74** [comment]
I work in a chemical plant and we use automation plc's.
These kind of hardware have a network cable and allows me to connect via
ftp to download the process data every day.
I am forced to set passive as false otherwise it does not connect...

ftplib has passive(False)

regards


2018-02-07 12:36 GMT+00:00 Nikita Melentev <notifications@github.com>:

**Comment by pohmelie** [comment]
Do you know the reason why `PASV` do not work? How it looks from client side?

**Comment by ramstein74** [comment]
for this test code:

import asyncio
import aioftp


async def get_file(host, port, login, password):
    async with aioftp.ClientSession(host, port, login, password) as client:
        for path, info in (await client.list(recursive=True)):
            if info["type"] == "file" and path.suffix == ".csv":
                await client.download(path)


loop = asyncio.new_event_loop()
tasks = (
    get_file("the_ip", 21, "user", "pw"),

)
loop.run_until_complete(asyncio.wait(tasks))
loop.close()


i get this error

Task exception was never retrieved
future: <Task finished coro=<get_file() done, defined at
<ipython-input-1-19f40198515f>:5> exception=StatusCodeError("Waiting
for ('227',) but got 502 [' PASV command not implemented.']",)>
Traceback (most recent call last):
  File "C:\Users\Ramos\AppData\Local\conda\conda\envs\app1\lib\site-packages\aioftp\client.py",
line 941, in get_passive_connection
    code, info = await self.command("EPSV", "229")
  File "C:\Users\Ramos\AppData\Local\conda\conda\envs\app1\lib\site-packages\aioftp\client.py",
line 248, in command
    self.check_codes(expected_codes, code, info)
  File "C:\Users\Ramos\AppData\Local\conda\conda\envs\app1\lib\site-packages\aioftp\client.py",
line 214, in check_codes
    raise errors.StatusCodeError(expected_codes, received_code, info)
aioftp.errors.StatusCodeError: Waiting for ('229',) but got 500 ["
'EPSV': command not understood."]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<ipython-input-1-19f40198515f>", line 7, in get_file
    for path, info in (await client.list(recursive=True)):
  File "C:\Users\Ramos\AppData\Local\conda\conda\envs\app1\lib\site-packages\aioftp\common.py",
line 120, in _to_list
    async for item in self:
  File "C:\Users\Ramos\AppData\Local\conda\conda\envs\app1\lib\site-packages\aioftp\client.py",
line 618, in __aiter__
    cls.stream = await cls._new_stream(path)
  File "C:\Users\Ramos\AppData\Local\conda\conda\envs\app1\lib\site-packages\aioftp\client.py",
line 615, in _new_stream
    return await self.get_stream(command, "1xx")
  File "C:\Users\Ramos\AppData\Local\conda\conda\envs\app1\lib\site-packages\aioftp\client.py",
line 978, in get_stream
    reader, writer = await self.get_passive_connection(conn_type)
  File "C:\Users\Ramos\AppData\Local\conda\conda\envs\app1\lib\site-packages\aioftp\client.py",
line 948, in get_passive_connection
    code, info = await self.command("PASV", "227")
  File "C:\Users\Ramos\AppData\Local\conda\conda\envs\app1\lib\site-packages\aioftp\client.py",
line 248, in command
    self.check_codes(expected_codes, code, info)
  File "C:\Users\Ramos\AppData\Local\conda\conda\envs\app1\lib\site-packages\aioftp\client.py",
line 214, in check_codes
    raise errors.StatusCodeError(expected_codes, received_code, info)
aioftp.errors.StatusCodeError: Waiting for ('227',) but got 502 ['
PASV command not implemented.']



2018-02-07 13:47 GMT+00:00 Nikita Melentev <notifications@github.com>:

**Comment by pohmelie** [comment]
Ok, so we can see both `EPSV` and `PASV` are not implemented. Very uncommon case, honestly. There is nothing to do with `aioftp` to make it work right now. I suggest you to use `ftplib`, since implement `PORT` for `aioftp` is long road.

**Comment by ramstein74** [comment]
thank u :(

2018-02-07 14:01 GMT+00:00 Nikita Melentev <notifications@github.com>:

**Pull request closed without merging**